rita_exit/src/database/database_tools.rs

```diff
@@ -348,11 +379,216 @@ pub fn create_or_update_user_record(
 
         let new_ip = get_next_client_ip(conn)?;
 
-        let c = client_to_new_db_client(client, new_ip, user_country);
+        let internet_ip = get_client_subnet(conn)?;
+
+        let c = client_to_new_db_client(client, new_ip, user_country, internet_ip);
 
         info!("Inserting new client {}", client.global.wg_public_key);
         diesel::insert_into(clients).values(&c).execute(conn)?;
 
         Ok(c)
     }
 }
+
+/// This function finds an available ipv6 subnet for a client that connects. It works as follows:
+/// 1.) Find assigned subnet and client configured subnet length (currently hardcoded to CLIENT_SUBNET_LENGTH)
+/// 2.) Retreive all active subnets from database
+/// 3.) Look at lazy static for any subnets in stack (This stack get populated when a client get deleted)
+/// 4.) If not, get our iterative index 'i' and generate the 'ith' client subnet out of the larger subnet
+/// Max Iterative index should theoretically never be reached because we choose subnets from deleted clients before generating
+/// an iterative subnet
+pub fn get_client_subnet(conn: &PgConnection) -> Result<IpNetwork, RitaExitError> {
+    // Get our assigned subnet
+    let rita_exit = settings::get_rita_exit();
+    let exit_settings = rita_exit.exit_network;
+    let subnet = exit_settings.subnet;
+    info!("Received Exit assigned ipv6 subnet: {}", subnet);
+
+    // Make sql query to get list of all client subnets in use
+    // SELECT internet_ipv6 FROM <TABLE>
+    use self::schema::clients::dsl::{clients, internet_ipv6};
+    let filtered_list = clients.select(internet_ipv6);
+    let ip_list = filtered_list.load::<String>(conn)?;
+
+    // First check for any available subnets to reclaim
+    let ip_tracker = &mut *CLIENT_IP_TRACKER.write().unwrap();
+    loop {
+        let addr = ip_tracker.available_subnets.pop();
+        if addr.is_none() {
+            // No free reclaimable subnets
+            break;
+        }
+        let addr = addr.unwrap();
+        if !ip_list.contains(&addr.to_string()) {
+            info!("Reclaimed an ip subnet! {:?}", addr);
+            return Ok(addr);
+        }
+    }
+
+    // Next get iterative index to generate next subnet. The reason we loop here is that there is a possiblity that a different exit instance has
+    // the same subnet as us and has already assigned some iterative ip addrs that we have not
+    for _ in 0..LOOP_THRESH {
```

Comment:
There's no database access in this loop, so I'm not sure how it prevents the case described in the comment. There's still a race condition that could occur while this loop is running so we would want to do a database check before and after allocation to prevent collision.

But like I said before I don't think we need to worry about this too much.

Reply:
Changed the logic to only pull from a new database table, removing the need for iteration anywhere.
1.) Retrieve subnet struct from database 
2.) Use reclaimed index from deleted client, else choose iterative index
3.) Use index to generate a subnet and update all info in databases

One thing I havent tested is whether my database queries are valid or not